Lane Arrows GUI in the Ground Plane using Unity Canvas

kvakvs opened the pull request:
* Shows GUI slightly above the ground
* Rotates the GUI accordingly to the segment orientation
* Respects clicks to GUI and updates GUI graphics
* Actually toggles the lanes
* TODO: Inherently supports Disabled lanes graphics but i've no idea how to detect a disabled lane
* TODO: Shader for see through (i have the shader just not loaded)

aubergine10 commented:
The alignment should ideally be determined a small distance from the node, rather than the angle the road connects to the node. In the second image in OP it's too rotated (from user perspective) IMO.

Would it be possible to just have one button, lane width, over the lane? Then left / right click would cycle through arrows in that direction? For example: A lane is currently right turn only. Left click makes it forward + right. Another left click makes it forward + left. Another left click makes it left turn only. A right click makes it forward + left...

In fact, the button could be invisible as the lane arrow on road will update (and, at later date, I suspect we'll also have lane connector lines visible while hovering a lane arrow as additional source of visual feedback + vital for roads that don't have lane arrow decals).

kvakvs commented:
Ok let me explain my flow of thought.

* Since we're editing that end of the segment and not the node, the arrows are displayed on the road and the GUI is rotated in the same direction.
* Having ground plane GUI stand up and be aligned to the projected screen like it's normal GUI or AR (augmented reality) GUI maybe is not a terrible idea but then we lose the benefit of orientation. I rather see the button rotation as a usability improvement.
* Having user click up to 8 times (2 to the power of 3 directions) to set up the lane is not a good UX. But having a turn sign and 3 smaller buttons (forward, right, left) around it might work.
  * Like so ![image](https://user-images.githubusercontent.com/288724/60081696-96763900-9732-11e9-8bca-aa79c23d0d6e.png)
* Reducing the amount of GUI and showing small buttons 1 per lane which then expand into a direction editor is again adding extra clicks, which might be a good UX or not good. I don't like the amount of clicks it takes to edit a segment in this case.
* If we go lane connector way, this is too easy to do wrong with terrible UX taking user dozens of clicks to set up.

aubergine10 commented:
Looking at the last image in comment above seems like that would be ideal situation - only one extra click to select a lane, then the three (or fewer, depending on what is possible at junction) buttons appear to set the arrow for that junction. IMO that would be better than showing all the buttons at once, for following reasons:

* Some roads have insane number of lanes or strange layouts, for example:
    * https://steamcommunity.com/sharedfiles/filedetails/?id=1565832628
    * https://steamcommunity.com/sharedfiles/filedetails/?id=1270405359
    * There are [more insane roads to test](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Roads-to-Test) in the wiki
* In some cases there is lots of stuff happening at a junction, for example:
    * Detailing props which might cover the buttons
    * Particularly things like concrete supports for overhead road would likely completely cover buttons if they are not entirely on a road

With the 'show all buttons' approach, it's taking up too much space on the ground increasing the risk that some of those buttons get covered, and for insanely large roads or weird layouts it's going to be confusing working out which lane buttons relate to which lane.

In terms of finding out what directions are possible at a junction, check the Junction Restrictions tool - I think there's some stuff in there (like the turn on red feature in particular) that has to work out if traffic can go left/right depending on driving direction and other roads connected to segment.

kvakvs commented:
I will re-consider this when i return from work.
One thing to note: I will be using a shader which ignores Z and will render through the world geometry.

aubergine10 commented:
The current lane connector renders through world geometry and that can still sometimes be difficult to use, for example, try lane connecting a node that's directly below a large elevated road. Luckily, with the lane arrow feature it only occurs at junctions which generally don't have stuff over the top of them.

One other option I considered some time ago was a feature that could flatten nearby props/buildings when working on a road, making it easier for the user to see what they are doing. No idea if that is possible though and suspect it would be difficult to implement.

kvakvs commented:
With the correct shader it will become a 3d overlay ignoring the world under it.
We do not need to modify any world geometry for it to work, it will be always visible and always clickable.

kvakvs commented:
These will be small buttons one per lane
Click on a small button will unfold 3 more control buttons ⬅⬆➡

![image](https://user-images.githubusercontent.com/288724/60134028-e559b800-979e-11e9-9dc3-875eef66acb3.png)

aubergine10 commented:
This is exactly the sort of direction I hoped lane arrows would go in. Very much looking forward to testing this once it progresses a bit.

kvakvs commented:
There you have it for testing
TODO: It prints error sometimes, there's some check failing in the lane change code (the game)
TODO: Some rare segments invert the lane arrows orientation by 180°, might need correction here

![image](https://user-images.githubusercontent.com/288724/60143737-0252b300-97c0-11e9-98de-d85fa3054269.png)

aubergine10 commented:
Regarding the 180° inversion, I think that's a standard thing with segments; they alternate their orientation so that any slight different in road texture edges doesn't create striping pattern. No idea how you would check for that though. There's probably some info on https://cslmodding.info in the networks section.

aubergine10 commented:
Looks like this PR will fix #368

kvakvs commented:
Current lane is highlighted
Changed button graphics
Added detection for impossible turns

![image](https://user-images.githubusercontent.com/288724/60220783-874cd380-9877-11e9-9619-ee833f2ff1d1.png)

aubergine10 commented:
This is starting to look really great!

When a lane is selected, is it possible to highlight connected outgoing lanes with green 'sausage' outline? That would be great visual feedback showing players exactly which outgoing lanes the current lane connects to.

The lane highlighting feature will be super useful for adding to several other TMPE features in future PRs.

krzychu124 commented:
@kvakvs 
I am using this approach to load shaders. Here is the code I use.
I've made simple Unity Editor extension:
(drop file to /Assets/Editor folder to get it working - you should see Example menu in Unity Editor)
```cs
public class Bundler : MonoBehaviour {
#if UNITY_EDITOR
    [UnityEditor.MenuItem("Example/Build Asset Bundles Using BuildMap")]
    static void BuildMapABs() {
        UnityEditor.AssetBundleBuild[] buildMap = new UnityEditor.AssetBundleBuild[1];
        buildMap[0].assetBundleName = "CustomShaders"; //bundle name, obviously
        string[] shaderAssets = new string[1];
        shaderAssets[0] = "Assets/CustomShaders/CustomShaderMin.shader";// path to asset source
        buildMap[0].assetNames = shaderAssets;
        UnityEditor.BuildPipeline.BuildAssetBundles("Assets/ShaderBundle", buildMap, 
        UnityEditor.BuildAssetBundleOptions.UncompressedAssetBundle, 
        UnityEditor.BuildTarget.StandaloneWindows64);
    }
#endif
}
```
Then in the mod I use this code to load from dll mod resources:
```cs
private void LoadShapeShaderBundle() {
    var myAssembly = System.Reflection.Assembly.GetExecutingAssembly();
    var myStream = myAssembly.GetManifestResourceStream("TrafficManager.Resources.shaders.customshaders");//added asset as Embedded resource of mod
    byte[] shader = TextureResources.ReadToEnd(myStream); //changed access from private to public
    AssetBundle bundle = AssetBundle.LoadFromMemory(shader);
    Shader sh = bundle.LoadAsset<Shader>("Assets/ShapeShader/CustomShaderMin.shader");//you can read that path to asset from bundle manifest but I hardcoded that because I am lazy :-)
//from now shader should be accessible via Shader.Find("Internal-Shader-Name"); (didn't tested)
    Material newMaterial = new Material(sh); //new material with shader
}
```

kvakvs commented:
@krzychu124 Let's see if the Unity editor can be removed from the build, ideally i want to press build in VS and have it done, not having to visit Unity.

krzychu124 commented:
Ok, but this is one time operation, I mean you have to just export shader as asset bundle and then you can even uninstall Unity editor ;) 
That's the only way I found to load external assets at runtime.

I am still testing new shader so it's way faster to write scripts directly in Unity and do some tests before migrating to mod ;)

kvakvs commented:
OK let's say I am happy to use bundle, if it is possible to embed that C# bundle call into a build script and run from Visual Studio, or have a standalone import tool that will be called from Visual Studio. Because we can store more in it than just 1 shader, like everything even.

krzychu124 commented:
So in other words, you want an ability to create asset bundle each time from scratch when you build mod dlls to be able to alter bundle content without involving Unity to create that bundle?

I will dig for more info about loading external assets and I will try to create automated script to bundle whole content of `assets` folder which we can create for shaders and more new things, like you've said :)

kvakvs commented:
Added outgoing lanes display
Added ESC handling on the tool window closing

![image](https://user-images.githubusercontent.com/288724/60309724-943cf600-994f-11e9-942a-38909378b686.png)

kvakvs commented:
Praise holy gods of geometry, i've solved inverted road segments.
I mean i detect them and invert the GUI correctly.

aubergine10 commented:
Some UX feedback based on quick test of this PR:

* Now that buttons are drawn on ground, I think it might look better if disabled buttons were removed (eg. I can't turn left, currently I see a left button with a strike through it) - only show the buttons I can use.
* After clicking a lane button (which shows the additional direction buttons in front of it), it would be nice if I could right-click to deslect the current lane. (I know I don't need to, I can just click the button for another lane, but as it's a common pattern for other tools it would be nice to include it here also)
* Rather than selecting a segment to customise, it would be nicer to select a node and then I can just work my way around the node lane by lane
* It would be nice if all "linked" outgoing lanes for a node were subtly sausage-highlighted. That way I can see instantly if any outgoing lanes are inaccessible at the node.
* If an incoming lane doesn't go to an outgoing lane, highlight that incoming lane as red? Outgoing lanes with no incoming lane could also be highlighted red?
* When hovering over a lane button (the one that toggles the direction buttons) it would be nice if it's currently "linked" outgoing lanes could be highlighted more strongly = gives me a way to see what outgoing lane(s) relate to an incoming lane just by hovering the incoming lane. Also nice would be to hover an outgoing lane and see the incoming lane(s) feed it.

aubergine10 commented:
The lane arrows that are part of road don't seem to be updating btw:

![image](https://user-images.githubusercontent.com/1386719/60557270-c928be80-9d3c-11e9-9132-1be6ae3ae96c.png)

Note the lane that can go forward and right, but the decal on the road only points forward.

kvakvs commented:
@aubergine10 So you want a whole new edit mode for nodes
As for hovering and highlighting i will re-read it later and try add to this graph as well
![Uploading TMPE_lanearrows.png…]()

aubergine10 commented:
I also ran in to some bugs last night but my internet crashed. I think possibly you are already aware of the bug but just in case: After taking the screenshot I posted in my previous comment, I clicked on the "left only" lane button and got following error:

```
Tool error: Argument is out of range.
Parameter name: index
  at System.Collections.Generic.List`1[UnityEngine.UI.Graphic].get_Item (Int32 index) [0x00000] in <filename unknown>:0 
  at UnityEngine.UI.GraphicRaycaster.Raycast (UnityEngine.EventSystems.PointerEventData eventData, System.Collections.Generic.List`1 resultAppendList) [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.CanvasGUI.WorldSpaceGUI.RaycastMouse () [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.SubTools.LaneArrows.LaneArrowTool.IsCursorInPanel () [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.TrafficManagerTool.determineHoveredElements () [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.TrafficManagerTool.SimulationStep () [0x00000] in <filename unknown>:0 
  at ToolManager.SimulationStepImpl (Int32 subStep) [0x00000] in <filename unknown>:0   [Core]
```

The error seems transient, I had performed same action multiple times prior to that with no error and I've not been able to recreate it since. [output_log.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3355186/output_log.txt)

I noticed there was no mention of the error in [TMPE.log](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3355187/TMPE.log) - ideally all TMPE errors should be mentioned in that log as well as main game log. TMPE has it's own logging util for simple string messages, so, if possible, maybe have something like:

```csharp
try {
   // ...
} catch (Exception ex) {
   Log.Warning($"An {ex.GetType().Name} occurred while using {ToolName}: See main game log for details"); // TMPE logger
   Debug.LogException(ex); // Unity logger
}
```

aubergine10 commented:
Regarding your flowchart, I was also thinking of this: #408

kvakvs commented:
Now i want to see a plan where this ticket can be considered good enough and delivered, and where #408 will continue. Without doing too much unneeded work.
I'd suggest to leave node select mode for #408, or rather as a sub-task to #408, it will reuse the same existing code in here.

aubergine10 commented:
For #391, I suggest this plan to bring to completion (most if not all of this will be retained should we later do #408): https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/391#issuecomment-507904974

kvakvs commented:
![image](https://user-images.githubusercontent.com/288724/60632313-3d2e9980-9e05-11e9-8977-09aa41a2e39f.png)

![image](https://user-images.githubusercontent.com/288724/60632337-50da0000-9e05-11e9-9ec5-f3c2e057dd61.png)

aubergine10 commented:
If an arrow is not available, maybe collapse and center the remaining arrows (2nd image)?

As for the overlapping, yeah, iswym. Can the lane buttons be moved on to the actual road segments, rather than being inside the junction node?

kvakvs commented:
I think they can be moved to more convenient places. For now it is too much noise on the junction.
We can see how to place the buttons better, tomorrow after work.

aubergine10 commented:
Do the lane arrows need the white backgrounds?

Another alternative could be if users were able click the actual road lanes (show sausage highlight on hover of a lane) and maybe only display our arrows if the road itself is detected not to have its own arrow decals?

kvakvs commented:
> Do the lane arrows need the white backgrounds?

A button has to look like a button. If i remove the background they will be barely visible and floating arrows won't look like buttons. I will investigate how to get rid of them, the lane mouse aiming is what we need, do we have code for that.

> Another alternative could be if users were able click the actual road lanes (show sausage highlight on hover of a lane) and maybe only display our arrows if the road itself is detected not to have its own arrow decals?

is there a way to know whether decals are visible, from the NetInfo?

aubergine10 commented:
> the lane mouse aiming is what we need, do we have code for that.

Not that I've seen, but @krzychu124 might know better.

> is there a way to know whether decals are visible, from the NetInfo?

I assume so, as there are some mods that can hide the lane arrows. _searches some git repos..._

Yeah, they are lane props apparently, some examples from mods that hide them:

* [Hide It](https://github.com/keallu/CSL-HideIt/blob/master/HideIt/Helpers/NetPropsHelper.cs#L11)
* [Remove Street Arrows](https://gist.github.com/simon56modder/161f68433ad951b3f6b10ad7c392d277)

aubergine10 commented:
Also the lane props likely use lane flags to work out what arrow to display, which would likely be more reliable way to detect them (in case their prop naming can be changed by asset creators): https://cslmodding.info/asset/network/#lane-flag-forward

aubergine10 commented:
If a road doesn't have it's own arrows, we could potentially re-use the vanilla road arrow textures when adding our own arrows to the road during lane arrow editing mode.

kvakvs commented:
This is the idea
Button use is to be minimised (maybe we still can use a button here and there), like ❌ for click-cancel

![TMPE_new_lanearrows](https://user-images.githubusercontent.com/288724/60655398-ef8a4f00-9e4d-11e9-8181-01aeedf7bb43.png)

aubergine10 commented:
Yes, yes, and thrice yes.

aubergine10 commented:
Although....

Should right-click in node select mode exit the lane arrow tool? That might become frustrating due to accidental clicks?

In arrow mode, in addition to being able to right-click back to lane select mode, would it be possible to add the ability to just click on any other incoming lane to switch arrow mode to that lane?

Likewise, in any mode, I should be able to click another node on the map to select it, without needing to right-click out of the current node or lane.

kvakvs commented:
Updated above

kvakvs commented:
@aubergine10 it feels like lane connections might become just another mode for junction lane editor described above. Instead of clicking on outgoing roads to enable, you'd click outgoing lanes with exact same UI and state transitions.
I believe the direction edit mode will not need any buttons whatsoever, maybe except the ❌ cancel/back

![image](https://user-images.githubusercontent.com/288724/60676114-5b84ab80-9e7e-11e9-8312-b9dd0a623ea0.png)

aubergine10 commented:
I had long hoped that these too tools could be merged somehow, but never expected we'd be able to progress it this quickly. The 'sausages' (lane highlighting) and ability to click actual lanes is the game changer.

Need some time to ponder, will post thoughts after dinner.

aubergine10 commented:
Screw dinner. Brain in overdrive. I'm actually starting to think all the node related features can go in a single tool. God-tier scope creep incoming... if I can work out how to put what's in my brain in to words... need a brew, brb

aubergine10 commented:
> ⚠️ This comment  focuses on _junction nodes_ - it doesn't account [other types of node](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Nodes,-Segments,-Lanes).

So, lane arrows and connectors are essentially two ways to achieving same task.

* Arrows = macroscopic: "connect me to lanes in that general direction"
* Connectors = microscopic: "connect me to these specific lanes"
* Arrows can't _currently_ be used on non-junction nodes, whereas connectors can

From gamification perspective, the connectors are really good:

* Partly due to the intricacy of clicking small circles
* The feeling of "wiring lanes together" is very rewarding

Here's my current thoughts...

At very least, we should merge these two tools (arrows, connectors) as they tackle the same problem.

* For **microscopic "connector" mode**, we definitely choose _outgoing lanes_, and they can be in any direction.
    * While making connections, the "dragged lane connector line" could be visible (like how current lane connector has the line between source lane and mouse pointer)
* For **macroscopic "arrow" mode**, we're not choosing _outgoing lanes_, we're choosing _outgoing directions_. So for that I think we should be clicking _outgoing segments_, not _outgoing lanes_.

We'd need some way to toggle between macroscopic vs. microscopic. An existing UX trope can be found in speed limits tool, where there's a checkbox to toggle, and user can also press a modifier key to temporarily switch mode.

It would be nice to see the lane connection lines in both modes (reasoning for this in a moment):

* Arrow-based connections should be thin white (dashed?) lines between incoming lane and, based on arrow direction, the resulting outgoing lanes
* Connector-based connections would be slightly thicker & coloured lines between incoming lane and the user-designated outgoing lanes
* The small circles would be removed, as now we have lane sausages & lane clicking
* The small arrows, depicting applicable outgoing lanes, could potentially be retained?
    * These were WIP on the lane connector PR, essentially they would appear when a lane is selected to illustrate more clearly what it could be connected too; also a filled arrow would indicate "at least something connects here"
* Could possibly toggle connection line visibility between: All, Current lane, None.

With lane connection lines, there are some pros and cons:

* They can look cluttered, especially on complex junctions (visibility modes mentioned above can help here)
* Regardless of which incoming lane is currently selected, the connection lines show _exactly_ what all the other lanes connect to/from
* We can display conflict points along those lane connection lines - see: #4 

If we are showing conflict points, we can show effect of priority signs on those conflicts. For example, if I make this incoming lane "Yield" or "Stop", how does that affect conflict points?

I'll leave it there for now, as there's likely issues in my thinking above.

kvakvs commented:
For macroscopic inputs I can take hover over segments, and treat it as a general direction. I.e. mouse pointer provides enough info for me to figure out whether it is left, forward or right

![image](https://user-images.githubusercontent.com/288724/60687329-31e37880-9eae-11e9-9916-d580eefde385.png)

aubergine10 commented:
I guess with macroscopic stuff, there is chance it's not just clicked outgoing lane/segment that will be connected, but could be several as per your diagram above? So if I click something going left, but there's a bunch of other segments that are "left of the incoming lane" then they'd all go green. This new UI is a superb enhancement over the previous UI in which I had no way to tell what a given arrow direction really connects to.

kvakvs commented:
Yes you got it right. But not only you will click it, i will render all affected lanes before you click it, so it should be kind of responsive and visual. Will have to extend tool base class a bit, now it handles nodes and segments, then it will also handle lanes.

aubergine10 commented:
Sweet, sort of like a preview "if you click this, here's what will happen" type thing? Different color for the preview (eg. green = current outgoing lanes, yellow = will be removed, blue = will be added)?

kvakvs commented:
As I control sausage width and transparency, I've got more wiggle room to create some sort of highlight without having to invent new colors. A wider translucent white outline might already be enough to show which lanes are targeted. The click will toggle them green or black/dark (i.e. off).

kvakvs commented:
How to display mixed state when some lanes have directions, and some have lane links?
If this mixed state is not allowed, how do we handle transitions?
- Lane had some links, and we want to convert it to directions?
- Lane had some directions and we want to add custom lane links?
- Allow both to coexist?
- Generalize to represent everything with lane links?

aubergine10 commented:
The routing manager, IIRC, is what handles the connections as far as pathfinder is concerned. It has two types of connection, one that the pathfinder must follow (ie. lane connections) and one that it can choose to ignore (lane arrows) if mod settings allow.

Victor explains in greater detail in this issue: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/368

kvakvs commented:
Let's do the general directions (lane arrows first), then this can be merged as a "complete feature replacing other similar feature", and then i can make another one for lane connectors taking colored curves and stuff in. They can be made a part of the same tool, and can coexist even without toggling options or holding Shift.

aubergine10 commented:
> One problem. I need some GUI space (a tooltip or a floating window?) to display hints, keyboard shortcuts and current help. A tooltip would be great, maybe.

I have some ideas for this, but will post them in an issue ticket where they can be discussed separately

kvakvs commented:
I have pushed a preview branch. It does no editing only view.
Enter modes by left clicking stuff. Leave modes by right clicking, pressing ESC or clicking another node.
@aubergine10

aubergine10 commented:
Looking good, some feedback...

When tool is activated, there's no indication of which nodes I can click until I hover the mouse over a node. For example, there are at least 5 nodes I can edit in this screenshot, but I have no visual cues to that effect:

> ![nodes](https://user-images.githubusercontent.com/1386719/60758011-aed73500-a009-11e9-8b46-61bb277cad99.jpg)
> If no node is currently selected, could the applicable nodes pulse so the user can clearly see what can be clicked?

Some transition nodes are perhaps not applicable to the lane arrow tool, for example there's not really anything I can do with this node:

> ![light-transition](https://user-images.githubusercontent.com/1386719/60758050-52284a00-a00a-11e9-8d50-34d6450e5be1.jpg)
> If the number of vehicle lanes is the same on either side of the transition node, treat it like a segment node?

Mouse out doesn't seem to be happening properly, for example in the following screenshot my mouse is over the Parking Restrictions button:

> ![mouse out](https://user-images.githubusercontent.com/1386719/60758274-776a8780-a00d-11e9-8662-8c4773cb61e5.jpg)
> The white lane should be blue, as the mouse is nowhere near it.

It would be great if, when hovering an incoming lane, its currently assigned outgoing lanes were highlighted on the map. That way I could quickly scan the lanes around a node just by hovering them to see what they connect to.

aubergine10 commented:
If I click a node (eg. see "parking restrictions" screenshot above), the node select circle remains, making it look like I can click it again. But it's already selected. Because of the large mouse 'hit areas' around lanes (and nodes) if I click the node as second time I'm actually selecting an incoming lane.

IMO it would be better, for selected node, to hide the node circle and prevent the node from being primary-clicked unless I exit out of that node first.

kvakvs commented:
@aubergine10 we should consider changing these overlays here, to some magic shader tech, like running arrow animation maybe? ![Running Arrows](https://dl.ledtronics.com/ImagesAll/web-arrow-mat.gif) Also it would be so much better with on-screen hints, i'll maybe i quickly slap together a simpler on-screen UI to display that until we get a good looking OSD

kvakvs commented:
Now allowed turns show as slow blinking green on hover (trying to keep it easy for the colorblind players by not using red and green simultaneously) and trying to keep it not too christmasy-tree-like everything is blinking. Maybe we could use some sprites at this point.

Added OnChangedHovered(Node|Segment|Lane) to the Main tool class, and using the lane hover change to set up the allowed turns structure.
![image](https://user-images.githubusercontent.com/288724/60766764-d16f5980-a0ae-11e9-9f3d-ddfcc96042cf.png)

aubergine10 commented:
> trying to keep it easy for the colorblind players

Hrm, that reminds me, I need to complete #287 - the idea is that we'd put a setting in mod options for choosing color palette. It would default to material style colors but users could select different palette based on their vision needs.

> Maybe we could use some sprites at this point.

IMO sprites will make it more christmasy-tree-like

kvakvs commented:
Also fixed lane hovering/unhovering.
Now mouse cursor has to hit the ground within 50m from the lane end, otherwise it clears the hover.

aubergine10 commented:
I still have some issues with hover, for example in the following image the mouse is over the blue sausage (just under the lamp post near the red letter box) but the node to the left is treated as being hovered:

![image](https://user-images.githubusercontent.com/1386719/60769990-24a4d480-a0ce-11e9-9d66-ec039448248b.png)

The current '50m from lane end' (and I suspect something similar for nodes) is too big margin for error.

For nodes, I think the mouse bust be over the node circle to treat it as hovered. For segments, I assume the issue is extra calculations required to deal with the fact they are curved?

Could we do something along these lines (pseudo-code):

```
function WhatAmIHovering() {

    for ( node in findNodesNearMouse() ) {
        if ( actuallyHovering( node ) ) return node
    }

    for ( lane in findLanesNearMouse() ) {
        if ( actuallyHovering( lane ) ) return lane
    }

    return null

}
```

And before calling that fictional function, check to see if what we were previously hovering is still being hovered (so as to avoid calling function unnecessarily).

----

Regarding the **green circles** that show current connections, I suggest remove the pulsing and leave them dimmed. I think ultimately those will need replacing with lane connection lines as it still took me a while to work out what they were showing, but they can be left for now as a placeholder (Krzychu is having some issues getting his new shader working in-game so might be a while before lane connection lines are ready for integration).

----

In terms of the pulsing effect, btw, is it possible to make it ease-in-ease-out? Think gently pulsing 'heartbeat' sleep/power indicator on Macbooks. The current pulsing seems to be a sawtooth pattern which feels like I'm getting visually slapped in the face over and over 🤣 

----

When I have a lane selected, and there are other lanes on that same segment:

![image](https://user-images.githubusercontent.com/1386719/60770249-ed83f280-a0d0-11e9-980e-e28a0cb64134.png)

It would be nice if I could click the other lanes to select them without first having to deselect the current lane.

aubergine10 commented:
If I select a node, then right click to exit that node, then right-click again to exit tool - it exits tool, but the lane arrows button is still shown as selected on the TM:PE menu bar.

kvakvs commented:
Have a try now, looking good.

Regarding switching editable lane, it means i need to render incoming sausages additionally, and they need to look different from outgoing. Maybe some shader with rolling colors or moving wave or something to show incoming.
https://www.androidpolice.com/wp-content/uploads/2016/01/nexus2cee_newanimationwaze4.gif

aubergine10 commented:
This is the first time I've felt in control of lane arrows tool. Instead of guessing what it's doing I can see exactly what it's doing. Once krzychu gets the other shader working, I think it will benefit from triangles on the outgoing lanes to more clearly differentiate incoming vs. outgoing. But it's definitely already a massive improvement on the original lane connector.

aubergine10 commented on TLM/TLM/UI/IncompatibleModsPanel.cs:
```diff
@@ -336,7 +336,11 @@ private void TryPopModal()
             if (blurEffect != null && UIView.ModalInputCount() == 0)
             {
-                ValueAnimator.Animate("ModalEffect", delegate (float val) { blurEffect.opacity = val; }, new AnimatedFloat(1f, 0f, 0.7f, EasingType.CubicEaseOut), delegate () { blurEffect.Hide(); });
+                ValueAnimator.Animate(
```
It would be better if stuff like this went in separate PRs

kvakvs replied:
I think it is doable to separate all the whitespace refactorings and minor formatting and the actual changes for lane arrows. Some of them will go entirely into lane arrows PR. Then can we get the whitespace/formatting merged first?

aubergine10 replied:
Yup, makes sense. Have to be careful the whitespace updates don't cause more merge conflicts with harmony branch though, that's already going to be nightmare to merge.

kvakvs replied:
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/419

krzychu124 replied:
I will merge master to harmony branch tomorrow ;)
